File: python/vectorized.py
import numpy as np
from math import sqrt
import numba as nb


def f(x):
    return 3 * (x ** 2) + 5 * x + 2


def vec(x):
    return f(2 * (x ** 2) + 6 * (x ** 3) - np.sqrt(x))


def devec(x):
    for i in range(len(x)):
        x_bar = x[i]
        x[i] = f(2 * x_bar ** 2 + 6 * x_bar ** 3 - sqrt(x_bar))
    return x


@nb.njit(cache=True)
def f_nb(x):
    return 3 * (x ** 2) + 5 * x + 2


@nb.njit(cache=True)
def devec_nb(x):
    x_new = np.zeros_like(x)
    for i in range(len(x)):
        x_new[i] = f_nb(2 * x[i] ** 2 + 6 * x[i] ** 3 - sqrt(x[i]))
    return x_new


x = np.abs(np.random.randn(1000000))

File: julia/vectorized.jl
f(x) = 3x.^2 .+ 5x .+ 2

vec(x) = f(2x.^2 + 6x.^3 - sqrt.(x))

newvec(x) = f.(2 .*x.^2 .+ 6 .* x.^3 .- sqrt.(x))

function devec(x)
    x_new  = x[:]
    @simd for i in eachindex(x)
        ϵ = x[i]
        x_new[i] = f(ϵ^2 + ϵ^3 - sqrt(ϵ))
    end
    return x
end

using BenchmarkTools
BenchmarkTools.DEFAULT_PARAMETERS.seconds = 10

x = abs.(randn(10^6))

t_vec = @benchmark devec(x)